commonmark/src/main/java/org/commonmark/parser/InlineParser.java

```diff
@@ -14,4 +13,9 @@
      */
     void parse(String input, Node node);
 
+    /**
+     * @return how many characters were parsed as a reference, {@code 0} if none
+     */
+    int parseReference(String s);
```

Comment:
I'm not sure if this belongs here or not, because it forces implementations of `InlineParser` to keep the reference map as state and it bakes this ugly part of the parser into the API.

For another issue, we'd also like to preserve parsed link reference definitions as nodes in the AST so that they can be used to render the original Markdown source back. So we'd have to break this API if we wanted to implement that.

Maybe I can look into that soon, but in the meantime, is there a way for you to not add this to the API while still enabling you to customize inline parsing?